CONTRIBUTING.md

```diff
@@ -111,73 +143,80 @@ Make sure you document each commit and squash them appropriately. You can use th
 * Scala's documentation on [Git Hygiene](https://github.com/scala/scala/tree/v2.12.0-M3#git-hygiene)
 * Play's documentation on [Working with Git](https://www.playframework.com/documentation/2.4.4/WorkingWithGit#Squashing-commits)
 
-Documentation
--------------
-
-Documentation fixes and contributions are as much welcome as to the source code itself. Visit [the website project](https://github.com/sbt/website) to learn about how to contribute.
-
 Build from source
 =================
 
-1. Install the current stable binary release of sbt (see [Setup]), which will be used to build sbt from source.
-2. Get the source code.
-
-		$ git clone git://github.com/sbt/sbt.git
-		$ cd sbt
+### Branch to work against
 
-3. The default branch is the development branch [1.0.x](https://github.com/sbt/sbt/tree/1.0.x), which contains the latest code for the next major sbt release.  To build a specific release or commit, switch to the associated tag.  The tag for the latest stable release is [v0.13.13](https://github.com/sbt/sbt/tree/v0.13.13):
+sbt uses two branches for development:
 
-		$ git checkout v0.13.13
+- Development branch: `1.x` (this is also called "master")
+- Stable branch: `1.$MINOR.x`, where `$MINOR` is current minor version (e.g. `1.1.x` during 1.1.x series)
 
-	Note that sbt is always built with the previous stable release.  For example, the [1.0.x](https://github.com/sbt/sbt/tree/1.0.x) branch is built with 0.13.13 and the [v0.13.13](https://github.com/sbt/sbt/tree/v0.13.13) tag is built with 0.13.12.
+If you're working on a bug fix, it's a good idea to start with the `1.$MINOR.x` branch. Since we can always safely merge from stable to `1.x`, but not other way around.
 
-4. To build the launcher and publish all components locally,
+### Instruction to build all modules from source
 
-		$ sbt
-		> publishLocal
+1. Install the current stable binary release of sbt (see [Setup]), which will be used to build sbt from source.
+2. Get the source code.
 
-5. To use this locally built version of sbt, copy your stable `~/bin/sbt` script to `~/bin/xsbt` and change it to use the launcher jar at `<sbt>/launch/target/sbt-launch.jar`.
+   ```
+   $ mkdir sbt-modules
+   $ cd sbt-modules
+   $ for i in sbt io util librarymanagement zinc; do \
+     git clone https://github.com/sbt/$i.git && (cd $i; git checkout -b 1.1.x origin/1.1.x)
+   done
+   $ cd sbt
+   $ ./sbt-allsources.sh
+   ```
 
-	Directory `target` is removed by `clean` command. Second solution is using the artifact stored in the local ivy repository.
+3. To build the launcher and publish all components locally,
 
-	The launcher is located in:
+   ```
+   $ ./sbt-allsources.sh
+   > ;{../io}/publishLocal; {../util}/publishLocal; {../librarymanagement}/publishLocal; {../zinc}/publishLocal
```

Comment:
Also need to `publishBridgesAndSet 2.12.4` in the zinc project. I couldn't figure out how to do that via `sbt-allsources`.

Reply:
Can this documentation be replaced with a command alias in the build?

Reply:
I tried:

```
sbt:sbtRoot> set commands ++= (commands in ProjectRef(uri("../zinc"), "zincCore")).value
[info] Defining commands
[info] The new value will be used by no settings or tasks.
[info] Reapplying settings...
^[[A[info] Set current project to sbtRoot (in build file:/Users/jz/code/sbt-modules/sbt/)

sbt:sbtRoot> publishBridgesAndSet 2.12.4
[error] Expected ID character
[error] Not a valid command: compilerInterface
[error] Expected project ID
[error] Expected configuration
[error] Expected ':'
[error] Expected key
[error] Not a valid key: compilerInterface (similar: compilerCache, compileOrder, compileIncSetup)
[error] compilerInterface/publishLocal
[error]                  ^
```

Reply:
Maybe we should be the command definition in a packaged `.scala` file in the zinc build so we can access it from the SBT build under `sbt-all-sources`.

Reply:
Okay, so this seems to work in the absence of a working `+` for this use case.

```
set {val p=ProjectRef(uri("../zinc"), "compilerBridge");  scalaVersion in p := (crossScalaVersions in p).value(0)}
{../zinc}compilerBridge/publishLocal

set {val p=ProjectRef(uri("../zinc"), "compilerBridge");  scalaVersion in p := (crossScalaVersions in p).value(1)}
{../zinc}compilerBridge/publishLocal

set {val p=ProjectRef(uri("../zinc"), "compilerBridge");  scalaVersion in p := (crossScalaVersions in p).value(2)}
{../zinc}compilerBridge/publishLocal
```

Reply:
Here's what I was about to commit

```scala
  commands += Command.command("publishLocalAllModule") { state =>
    val extracted = Project.extract(state)
    import extracted._
    val sv      = get(scalaVersion)
    val projs   = structure.allProjectRefs
    val ioOpt   = projs find { case ProjectRef(_, id) => id == "ioRoot"; case _ => false }
    val utilOpt = projs find { case ProjectRef(_, id) => id == "utilRoot"; case _ => false }
    val lmOpt   = projs find { case ProjectRef(_, id) => id == "lmRoot"; case _ => false }
    val zincOpt = projs find { case ProjectRef(_, id) => id == "zincRoot"; case _ => false }
    (ioOpt   map { case ProjectRef(build, _) => "{" + build.toString + "}/publishLocal" }).toList :::
    (utilOpt map { case ProjectRef(build, _) => "{" + build.toString + "}/publishLocal" }).toList :::
    (lmOpt   map { case ProjectRef(build, _) => "{" + build.toString + "}/publishLocal" }).toList :::
    (zincOpt map { case ProjectRef(build, _) =>
      val zincSv = get(scalaVersion in ProjectRef(build, "zinc"))
      val csv = get(crossScalaVersions in ProjectRef(build, "compilerBridge")).toList
      (csv flatMap { bridgeSv =>
        s"++$bridgeSv" :: ("{" + build.toString + "}compilerBridge/publishLocal") :: Nil
      }) :::
      List(s"++$zincSv", "{" + build.toString + "}/publishLocal")
    }).getOrElse(Nil) :::
    List(s"++$sv", "publishLocal") :::
    state
  },
```

Reply:
With that, and the other instructions, I was able to modify `util` to beautify the logging and compile a sample project:

```
⚡ sbt
[❆info] ❆Loading settings from gpg.sbt,idea.sbt ...
[❆info] ❆Loading global plugins from /Users/jz/.sbt/1.0/plugins
[❆info] ❆Updating ProjectRef(uri("file:/Users/jz/.sbt/1.0/plugins/"), "global-plugins")...
[❆info] ❆Done updating.
[❆info] ❆Compiling 2 Scala sources to /Users/jz/.sbt/1.0/plugins/target/scala-2.12/sbt-1.0/classes ...
[❆info] ❆Done compiling.
[❆info] ❆Loading project definition from /private/tmp/sbt-scratch/project
[❆info] ❆Updating ProjectRef(uri("file:/private/tmp/sbt-scratch/project/"), "sbt-scratch-build")...
[❆info] ❆Done updating.
[❆info] ❆Loading settings from build.sbt ...
[❆info] ❆Set current project to sbt-scratch (in build file:/private/tmp/sbt-scratch/)
[❆info] ❆sbt server started at local:///Users/jz/.sbt/1.0/server/d783f1fc400c8f31587a/sock
sbt:sbt-scratch> compile
[❆info] ❆Compiling 1 Scala source to /private/tmp/sbt-scratch/target/scala-2.13.0-M2/classes ...
[❆info] ❆Non-compiled module 'compiler-bridge_2.13.0-M2' for Scala 2.13.0-M2. Compiling...
[❆info] ❆  Compilation completed in 7.275s.
[❆info] ❆Done compiling.
[❆success] ❆Total time: 8 s, completed 21/02/2018 5:21:32 PM
```